Test/ElasticScale.ShardManagement.UnitTests/ShardMapManagerUpgradeTests.cs

```diff
@@ -1,339 +1,448 @@
+                        Globals.ShardMapManagerConnectionString,
+                        ShardMapManagerLoadPolicy.Lazy);
+
+            // Upgrade to version 1.1
+            smm.UpgradeGlobalStore(new Version(1, 1));
+
+            // Create a range shard map and add few mappings
+            RangeShardMap<int> rsm = smm.CreateRangeShardMap<int>(ShardMapManagerUpgradeTests.s_shardMapNames[1]);
+            Assert.IsNotNull(rsm);
+
+            Shard s = rsm.CreateShard(new ShardLocation(Globals.ShardMapManagerTestsDatasourceName, ShardMapManagerUpgradeTests.s_shardedDBs[0]));
+            Assert.IsNotNull(s);
+
+            RangeMapping<int> m1 = rsm.CreateRangeMapping(new Range<int>(1, 10), s);
+            RangeMapping<int> m2 = rsm.CreateRangeMapping(new Range<int>(10, 20), s);
+            RangeMapping<int> m3 = rsm.CreateRangeMapping(new Range<int>(20, 30), s);
+
+            // Lock first 2 mappings with same lockownerid and third with a different lock owner id
+
+            MappingLockToken t1 = MappingLockToken.Create();
+            MappingLockToken t2 = MappingLockToken.Create();
+
+            rsm.LockMapping(m1, t1);
+            rsm.LockMapping(m2, t1);
+            rsm.LockMapping(m3, t2);
+
+            // now try to unlock using token t2. In store version 1.1 it will unlock all mappings
+            rsm.UnlockMapping(t2);
+
+            foreach (RangeMapping<int> m in rsm.GetMappings())
+            {
+                Assert.AreEqual(MappingLockToken.NoLock, rsm.GetMappingLockOwner(m));
+            }
+
+            // Now upgrade to version 1.2 and try same scenario above.
+            smm.UpgradeGlobalStore(new Version(1, 2));
+
+            rsm.LockMapping(m1, t1);
+            rsm.LockMapping(m2, t1);
+            rsm.LockMapping(m3, t2);
+
+            // Unlock using token t1. It should just unlock 2 mappings and leave last one locked.
+            rsm.UnlockMapping(t1);
+
+            Assert.AreEqual(MappingLockToken.NoLock, rsm.GetMappingLockOwner(rsm.GetMappingForKey(5)));
+            Assert.AreEqual(MappingLockToken.NoLock, rsm.GetMappingLockOwner(rsm.GetMappingForKey(15)));
+            Assert.AreEqual(t2, rsm.GetMappingLockOwner(rsm.GetMappingForKey(25)));
+
+            // Cleanup - Delete all mappings. shard will be removed in test cleanup.
+            rsm.UnlockMapping(t2);
+            RangeMappingUpdate ru = new RangeMappingUpdate();
+            ru.Status = MappingStatus.Offline;
+
+            foreach (RangeMapping<int> m in rsm.GetMappings())
+            {
+                rsm.DeleteMapping(rsm.UpdateMapping(m, ru));
+            }
+        }
+
+        [TestMethod]
+        public void TestSerializationFixInVersion1_3()
+        {
+            // Get shard map manager
+            ShardMapManager smm = ShardMapManagerFactory.GetSqlShardMapManager(
+                Globals.ShardMapManagerConnectionString,
+                ShardMapManagerLoadPolicy.Lazy);
+

```

Comment:
smm.UpgradeGlobalStore(new Version(1, 2)) before proceeding. That way we can verify that issue exists in v1.2 and then got fixed in v1.3. This particular fix is simpler, but if we ever have a regression 'in-cycle' then starting from fixed version will catch it correctly.